src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ThrowExceptionForHRTests.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Runtime.InteropServices.Tests
+{
+    public class ThrowExceptionForHRTests
+    {
+        [Theory]
+        [InlineData(unchecked((int)0x80020006))]
+        [InlineData(unchecked((int)0x80020101))]
+        [ActiveIssue(30866)]
+        public void ThrowExceptionForHR_NoErrorInfo_ReturnsValidException(int errorCode)
+        {
+            bool calledCatch = false;
+            try
+            {
+                Marshal.ThrowExceptionForHR(errorCode);
+            }
+            catch (Exception ex)
+            {
+                calledCatch = true;
+
+                Assert.IsType<COMException>(ex);
+                Assert.Equal(errorCode, ex.HResult);
+                Assert.Null(ex.InnerException);
+                Assert.Null(ex.HelpLink);
+                Assert.NotEmpty(ex.Message);
+                Assert.Equal("mscorlib", ex.Source);
+                Assert.Contains(nameof(ThrowExceptionForHR_NoErrorInfo_ReturnsValidException), ex.StackTrace);
+                Assert.Contains(nameof(Marshal.ThrowExceptionForHR), ex.TargetSite.Name);
+            }
+
+            Assert.True(calledCatch);
+        }
+
+        public static IEnumerable<object[]> ThrowExceptionForHR_ErrorInfo_TestData()
+        {
+            yield return new object[] { unchecked((int)0x80020006), IntPtr.Zero };
+            yield return new object[] { unchecked((int)0x80020101), IntPtr.Zero };
+            yield return new object[] { unchecked((int)0x80020006), (IntPtr)(-1) };
+            yield return new object[] { unchecked((int)0x80020101), (IntPtr)(-1) };
+        }
+
+        [Theory]
+        [MemberData(nameof(ThrowExceptionForHR_ErrorInfo_TestData))]
+        [ActiveIssue(30866)]
```

Comment:
30866 is fixed(or wokaround). You need to call another Marshal.GetExceptionForHR to clear IErrorInfo in that thread. see https://github.com/dotnet/corefx/pull/31136

Reply:
@luqunl I think this should be addressed when @hughbe merges - seea754a612eed6d16b52d7f75c602f71eb367760ae.

Reply:
Yes. I'm going to wait on solving merge conflicts in this PR till my open `Marshal` PRs are merged (https://github.com/dotnet/corefx/pulls/hughbe)